### PR TITLE
[FIX] sale_loyalty: factor in tax rounding method for discounts

### PR DESCRIPTION
**Pull request by lvsz**
Versions
--------
- 16.0+

Steps
-----
1. Set tax rounding method to global;
2. create a loyalty program offering 100% discount on orders;
3. create a SO with the following line subtotals & 15% tax:
    1. $5.60
    2. $8.92
    3. $44.91
    4. $217.26
    5. $2400.00
4. apply the 100% discount program.

Issue
-----
Total comes down to $0.01 instead of $0.00.

Cause
-----
To calculate the `discountable` amount, the `_discountable_order` method uses the `_compute_taxes` method from `account.tax` for each line, which doesn't check for global rounding method. This can introduce small rounding errors. When calculating the discount factor by dividing the `max_discount` (total amount of the order) by `discountable`, the result will be 0.999... instead of 1 if `discountable` is off by one cent.

Solution
--------
Use the `compute_all` method instead, which factors in the tax rounding method by using increased precision.

opw-3710365

### CODE REVIEW

**Comment by lvsz** [comment]
A variant of this issue also exists when rounding taxes by line, leading to a different result than the tax of the 100% discount line, but I'm not sure that can be fixed, as the calculations are actually correct there.

**Review by nle-odoo — APPROVED** [review]
looks good to me

**Review by Demesmaeker — APPROVED** [review]
Hello @lvsz 

Thanks for your work here, it indeed solves the problem. ^^
I only pushed some nitpickings, if you are okay, you can squash it and merge.

Have a lovely day

@robodoo delegate+

File: addons/sale_loyalty/models/sale_order.py [hunk 1]
```diff
@@ -165,15 +165,22 @@ def _discountable_order(self, reward):
...
+                partner=line.order_partner_id,
+            )
+            # To compute the discountable amount we get the subtotal and add
+            # non-fixed tax totals. This way fixed taxes will not be discounted
             taxes = line.tax_id.filtered(lambda t: t.amount_type != 'fixed')
-            discountable_per_tax[taxes] += totals['amount_untaxed']
+            discountable += tax_data['total_excluded'] + sum(
+                tax['amount'] for tax in tax_data['taxes'] if tax['id'] in taxes.ids
+            )
+            discountable_per_tax[taxes] += line.price_reduce - sum(
```
Comment (Feyensv) [inline addons/sale_loyalty/models/sale_order.py:180]: https://github.com/odoo/odoo/blob/d44c45234577d19a728e92368f224b0e0e59e853/addons/sale/models/sale_order_line.py#L152-L153

Comment (Demesmaeker) [reply addons/sale_loyalty/models/sale_order.py:180]: Indeed, my bad :facepalm:

Comment (lvsz) [reply addons/sale_loyalty/models/sale_order.py]: Is replacing it with something like this acceptable?
```diff
@@ -177,7 +177,8 @@ class SaleOrder(models.Model):
             discountable += tax_data['total_excluded'] + sum(
                 tax['amount'] for tax in tax_data['taxes'] if tax['id'] in taxes.ids
             )
-            discountable_per_tax[taxes] += line.price_reduce - sum(
+            line_price = line.price_unit * line.product_uom_qty * (1 - line.discount / 100)
+            discountable_per_tax[taxes] += line_price - sum(
                 tax['amount'] for tax in tax_data['taxes']
                 if tax['price_include'] and tax['id'] not in taxes.ids
             )

```

**Review by Feyensv — APPROVED** [review]
Since the tests from #146947 still work fine, lgtm :ok_hand: 

@robodoo delegate+

File: addons/sale_loyalty/models/sale_order.py [hunk 2]
```diff
@@ -165,15 +165,23 @@ def _discountable_order(self, reward):
...
+                partner=line.order_partner_id,
+            )
+            # To compute the discountable amount we get the subtotal and add
+            # non-fixed tax totals. This way fixed taxes will not be discounted
             taxes = line.tax_id.filtered(lambda t: t.amount_type != 'fixed')
-            discountable_per_tax[taxes] += totals['amount_untaxed']
+            discountable += tax_data['total_excluded'] + sum(
+                tax['amount'] for tax in tax_data['taxes'] if tax['id'] in taxes.ids
+            )
+            line_price = line.price_unit * line.product_uom_qty * (1 - line.discount / 100)
```
Comment (Feyensv) [inline addons/sale_loyalty/models/sale_order.py:180]:
```suggestion
            line_price = line.price_unit * line.product_uom_qty * (1 - (line.discount or 0.0) / 100)
```
last little change to be on the safe side, like it's done in `sale`.

Comment (lvsz) [reply addons/sale_loyalty/models/sale_order.py]: Sure, added the change for consistency, though the ORM already ensures `Float` fields always return 0.0 as the falsy value:
https://github.com/odoo/odoo/blob/990aae55b1a2b95765a848b6fc122bb69f768ce1/odoo/fields.py#L1538-L1550

**Comment by lvsz** [comment]
@robodoo r+

**Pull request closed without merging**